azure/services/subnets/spec.go

```diff
@@ -135,3 +115,35 @@ func (s *SubnetSpec) Parameters(ctx context.Context, existing interface{}) (para
 		SubnetPropertiesFormat: &subnetProperties,
 	}, nil
 }
+
+// shouldUpdate returns true if an existing subnet should be updated.
+func (s *SubnetSpec) shouldUpdate(existingSubnet network.Subnet) bool {
+	// No modifications for non-managed subnets
+	if !s.IsVNetManaged {
+		return false
+	}
+
+	// Update the subnet a NAT Gateway was added for backwards compatibility.
+	if s.NatGatewayName != "" && existingSubnet.SubnetPropertiesFormat.NatGateway == nil {
+		return true
+	}
+
+	// Update the subnet if the service endpoints changed.
+	if existingSubnet.ServiceEndpoints != nil || len(s.ServiceEndpoints) > 0 {
+		var existingServiceEndpoints []network.ServiceEndpointPropertiesFormat
+		if existingSubnet.ServiceEndpoints != nil {
+			for _, se := range *existingSubnet.ServiceEndpoints {
+				existingServiceEndpoints = append(existingServiceEndpoints, network.ServiceEndpointPropertiesFormat{Service: se.Service, Locations: se.Locations})
+			}
+		}
+		newServiceEndpoints := make([]network.ServiceEndpointPropertiesFormat, len(s.ServiceEndpoints))
+		for _, se := range s.ServiceEndpoints {
+			se := se
+			newServiceEndpoints = append(newServiceEndpoints, network.ServiceEndpointPropertiesFormat{Service: pointer.String(se.Service), Locations: &se.Locations})
+		}
+
+		diff := cmp.Diff(newServiceEndpoints, existingServiceEndpoints)
```

Comment:
Would it be worth it to try to log the diff somewhere? I think it would help with debugging/maintaining the unit test for this function since it'd be hard to tell what went wrong if we're just surfacing a `true` or `false` result.